internal/pkg/agent/application/upgrade/step_relink.go

```diff
@@ -23,7 +23,8 @@ func ChangeSymlink(ctx context.Context, log *logger.Logger, targetHash string) e
 	hashedDir := fmt.Sprintf("%s-%s", agentName, targetHash)
 
 	symlinkPath := filepath.Join(paths.Top(), agentName)
-	newPath := filepath.Join(paths.Top(), "data", hashedDir, agentName)
+
+	newPath := paths.BinaryPath(filepath.Join(paths.Top(), "data", hashedDir), agentName)
```

Comment:
If we forget `paths.BinaryPath` here (or elsewhere) to wrap this is there a test that fails? If not, is there a unit test we could write?

I am concerned the need for this won't be obvious to the next engineer to touch this code, and it could be forgotten or removed in the future.

Reply:
I see the same thing in service_darwin.go, so this seems to be necessary in lots of places now but I am not sure if anything enforces the need for it.

```go
func invokeCmd(topPath string) *exec.Cmd {
	homeExePath := paths.BinaryPath(topPath, agentName)
```

Reply:
Yeah currently the logic is scattered in different places, so we call these wrapper func in few different places. Relying on e2e tests to catch any issues more or less. 
It's the same issue with the code before this change actually, nothing is enforcing the right path to be used. 
That’s how I found all of the places, changing in on place and finding the next places where it breaks while running e2e upgrades. 
I'll address the comments when backporting to main branch.